src/ir/auth_logic/ast_string.h

```diff
@@ -48,58 +50,123 @@ std::string ToString(AstNodeVariantType Node) {
                 predicate.name(), "(", absl::StrJoin(predicate.args(), ", "),
                 ")");
           },
+          [](datalog::ArgumentType argumentType) {
```

Comment:
Should `ToString` be implemented using the AST visitor? (Let us discuss and agree what the right thing is. Don't make the change yet.)

Reply:
I don't think that is better because emitting strings is a function on the AST where the structure of the AST and the function performed cannot be usefully decoupled because emitting a string is a function that requires a very specific order (which may not be exactly a pre-order or post-order traversal). (I think an example of a function that would make better use of the visitor pattern would be something like computing the set of subexpressions in a program with a particular property.)

Reply:
SG. We can leave this as is.